File: greedy/week3_2_Reduce_Array_Size_to_The_Half/szzng.js
var minSetSize = function(arr) {
    const cnt = {}
    arr.forEach(v=> {
        if (v in cnt) cnt[v]+=1
        else cnt[v]=1
    })

    const frequency = Object.values(cnt).sort((a,b)=>b-a)
    const half_len = parseInt(arr.length / 2)
    let current_len = arr.length

    for (var i=0; i < frequency.length; i++){
        current_len -= frequency[i]
        if (current_len <= half_len) return i+1
    }
};

File: greedy/week3_2_Reduce_Array_Size_to_The_Half/szzng.py
# Runtime 555 ms Beats 90.72%
# Memory 31.8 MB Beats 73.9%

from typing import List


class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        cnt = {}
        for i in arr:
            cnt[i] = cnt[i] + 1 if i in cnt else 1

        frequency = sorted(list(cnt.values()), reverse=True)
        half_len = len(arr) // 2
        current_len = len(arr)

        for i, f in enumerate(frequency):
            current_len -= f
            if current_len <= half_len: return i + 1
